[nativert] Move TensorMeta to pytorch core

yiming0416 opened the pull request:
Summary:
Torch Native Runtime RFC: https://github.com/pytorch/rfcs/pull/72

This diff moves `TensorMeta.cpp` and `TensorMeta.h` to PyTorch core under `torch/nativert/graph/`

Existing `torch::_export::TensorMeta` in `torch/csrc/utils/generated_serialization_types.h` is auto-generated from the export serde schema and therefore only containing the most basic serializable types. We need the newly added `TensorMeta.cpp` to deserialize the metadata into a in-memory class with c10 types so that it can be consumed by the runtime later. 

Test Plan:

Added test under `test/cpp/nativert/test_tensor_meta.cpp`

Differential Revision: D73820548

albanD commented:
Can you add some details in the description on how this is different from export::TensorMeta and why we need this separate one?

albanD commented on test/cpp/nativert/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+)
+
+# We also build with UBSAN flag in build_asan.h
+if(USE_ASAN)
```
We actually run this under asan?
Or is this a CMake from somewhere else?

yiming0416 replied:
I copied this from `test/cpp/jit/CMakeLists.txt`

albanD replied:
Yes let's remove everything we don't need from this cmake

yiming0416 replied:
@albanD sure! removed in the latest version.

albanD commented on build_variables.bzl:
```diff
@@ -619,7 +624,7 @@ libtorch_lite_cmake_sources = sorted(
 )
 
-libtorch_cmake_sources = libtorch_core_sources + libtorch_core_jit_sources
+libtorch_cmake_sources = libtorch_core_sources + libtorch_core_jit_sources + libtorch_nativert_sources
```
Should we make the runtime a separate .so like other subsets of libtorch to improve dependency management?
In particular, do you expect you users to only link against your function (and transitively get the rest of libtorch). Or you expect them to use whole of libtorch?

yiming0416 replied:
cc @zhxchen17

zhxchen17 replied:
I think we expect users to use the whole of libtorch.

yiming0416 commented:
> Can you add some details in the description on how this is different from export::TensorMeta and why we need this separate one?

@albanD Added details in the PR summary

albanD commented on torch/nativert/graph/TensorMeta.cpp:
```diff
@@ -0,0 +1,142 @@
+      return c10::MemoryFormat::ChannelsLast3d;
+    case torch::_export::MemoryFormat::PreserveFormat:
+      return c10::MemoryFormat::Preserve;
+    default:
+      TORCH_CHECK(
+          false, "unknown memory format", static_cast<int>(memoryFormat));
+  }
+}
+
+c10::Layout convertJsonLayout(const torch::_export::Layout& layout) {
+  switch (layout) {
+    case torch::_export::Layout::Unknown:
+      TORCH_CHECK(false, "got unknown layout");
+    case torch::_export::Layout::SparseCoo:
+      // TODO is this the right translation
+      return c10::Layout::Sparse;
+    case torch::_export::Layout::SparseCsr:
+      return c10::Layout::SparseCsr;
+    case torch::_export::Layout::SparseCsc:
+      return c10::Layout::SparseCsc;
+    case torch::_export::Layout::SparseBsr:
+      return c10::Layout::SparseBsr;
+    case torch::_export::Layout::SparseBsc:
+      return c10::Layout::SparseBsc;
+    case torch::_export::Layout::_mkldnn:
+      return c10::Layout::Mkldnn;
+    case torch::_export::Layout::Strided:
+      return c10::Layout::Strided;
+    default:
+      TORCH_CHECK(false, "unknown layout", static_cast<int>(layout));
+  }
+}
+
+c10::Device convertJsonDevice(const torch::_export::Device& device) {
+  c10::Device d(device.get_type());
+  if (auto index = device.get_index()) {
+    d.set_index(*index);
+  }
+  return d;
+}
```
nit: Can we move these next to the corresponding _export struct ? They have unrelated to the nativert::TensorMeta right?

yiming0416 replied:
hmmm the header file for `torch::_export::` is auto-generated from a script and only contains the most basic utils for ser/de to/from json. So we might want to keep these conversion functions for the in-memory representation of the TensorMeta in this file.

albanD replied:
Ok, as long as we don't make them public from here, that sounds like a good tradeoff for now

albanD commented on test/cpp/nativert/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+# Build the cpp gtest binary containing the cpp-only tests.
+set(NATIVERT_TEST_SRCS
+  ${NATIVERT_TEST_ROOT}/test_tensor_meta.cpp
+  ${TORCH_ROOT}/torch/nativert/graph/TensorMeta.cpp
```
Why do you need this file? You already depend on all of libtorch.

yiming0416 replied:
@albanD If I don't include this file, CI failed when building `test_nativert`, complaining about undefined references to functions defined in `TensorMeta.cpp`.  For example: https://hud.pytorch.org/pr/pytorch/pytorch/152475#41555463764

Including this file will make CI pass. I am also confused about this but haven't figured out why. Did I miss anything for the build? cc @zhxchen17

albanD replied:
Are these functions supposed to be public API on libtorch? Or just internal tooling?
If they're just internal tooling, it is expected that you cannot find them on libtorch (as we only expose implementations for our public APIs)

zhxchen17 replied:
Yeah as @albanD said, we don't need this to be in the test target.

yiming0416 replied:
@albanD In the latest version I made them public by adding `TORCH_API` macro to the functions, because unit-tests are dynamically linked to `libtorch` so the macro is needed for symbol visibility. Now I am able to remove the `TensorMeta.cpp` from the CMake file.

albanD replied:
But do you actually want these to be public?
That means you won't be able to move or change them for the foreseeable future.

yiming0416 replied:
so we have two options:
1. make the functions public, no need to include the `TensorMeta.cpp` in the test CMake
2. don't make them public, but need to include the `TensorMeta.cpp` in the test CMake.
@albanD when you say "won't be able to move or change", doesn't that mean we cannot change the public APIs due to BC?

If it's just for unblocking the linking issue for newly added unit tests in GitHub CI. I feel like option2 might be better? Internal CI doesn't have symbol visibility issues even if we don't mark the functions public. Otherwise we need to make all functions being tested public and seems to be a big commitment for BC. WDYT? @zhxchen17

albanD replied:
> doesn't that mean we cannot change the public APIs due to BC?

That is correct.

Note that this will be a common issue you'll have I guess to test all the non-public APIs.

I also agree that not making them public sounds better.

albanD commented on test/cpp/nativert/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+
+# Build the cpp gtest binary containing the cpp-only tests.
+set(NATIVERT_TEST_SRCS
+  ${NATIVERT_TEST_ROOT}/test_tensor_meta.cpp
```
nit: Since I assume all the nativert tests will live here, can we please have some answer on when do they run and on which configs?

albanD reviewed (commented):
A couple nits, real question on the test cmake though !

yiming0416 commented:
@albanD  Can you pls review again? I have made APIs private and added the corresponding source file to the test cmake file. If there is any other suggesting changes before merging, I'd be happy to make those. Thanks!

albanD reviewed (approved):
Sounds ok!

yiming0416 committed:
[nativert] Move TensorMeta to pytorch core (#152475)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152475

Torch Native Runtime RFC: https://github.com/pytorch/rfcs/pull/72

Thi diff moves `TensorMeta.cpp` and `TensorMeta.h` to PyTorch core.

Test Plan:
Internal CI.
GitHub CI with newly added test under `test/cpp/nativert/test_tensor_meta.cpp`

Reviewed By: zhxchen17

Differential Revision: D73820548
